File: stepik-java-api/src/main/java/org/stepik/api/queries/metrics/StepikMetricsPostQuery.java
package org.stepik.api.queries.metrics;

import org.jetbrains.annotations.NotNull;
import org.stepik.api.actions.StepikAbstractAction;
import org.stepik.api.objects.metrics.MetricWrapper;
import org.stepik.api.queries.StepikAbstractPostQuery;
import org.stepik.api.queries.VoidResult;
import org.stepik.api.urls.Urls;

import java.util.Map;

/**
 * @author meanmail
 */
public class StepikMetricsPostQuery extends StepikAbstractPostQuery<VoidResult> {
    private final MetricWrapper metrics = new MetricWrapper();

    public StepikMetricsPostQuery(@NotNull StepikAbstractAction stepikAction) {
        super(stepikAction, VoidResult.class);
    }

    @NotNull
    public StepikMetricsPostQuery name(@NotNull String name) {
        metrics.getMetric().setName(name);
        return this;
    }

    @NotNull
    public StepikMetricsPostQuery timestamp(long value) {
        metrics.getMetric().setTimestamp(value);
        return this;
    }

    @NotNull
    public StepikMetricsPostQuery tags(@NotNull Map<String, String> value) {
        metrics.getMetric().setTags(value);
        return this;
    }

    @NotNull
    public StepikMetricsPostQuery tags(@NotNull String key, @NotNull String value) {
        metrics.getMetric().addTags(key, value);
        return this;
    }

    @NotNull
    public StepikMetricsPostQuery tags(@NotNull String key, @NotNull Object value) {
        metrics.getMetric().addTags(key, value.toString());
        return this;
    }

    @NotNull
    public StepikMetricsPostQuery data(@NotNull Map<String, Object> value) {
        metrics.getMetric().setData(value);
        return this;
    }

    @NotNull
    public StepikMetricsPostQuery data(@NotNull String key, @NotNull Object value) {
        metrics.getMetric().addData(key, value);
        return this;
    }

    @NotNull
    @Override
    protected String getBody() {
        return getJsonConverter().toJson(metrics);
    }

    @NotNull
    @Override
    protected String getUrl() {
        return Urls.METRICS;
    }

    @Override
    public String toString() {
        return getJsonConverter().toJson(metrics);
    }
}

File: stepik-java-api/src/main/kotlin/org/stepik/api/queries/metrics/StepikMetricsPostQuery.kt
package org.stepik.api.queries.metrics

import org.stepik.api.actions.StepikAbstractAction
import org.stepik.api.objects.metrics.MetricWrapper
import org.stepik.api.queries.StepikAbstractPostQuery
import org.stepik.api.queries.VoidResult

class StepikMetricsPostQuery(stepikAction: StepikAbstractAction) :
        StepikAbstractPostQuery<VoidResult>(stepikAction, VoidResult::class.java) {
    
    private val metrics = MetricWrapper()
    
    fun name(name: String): StepikMetricsPostQuery {
        metrics.getMetric()
                .name = name
        return this
    }
    
    fun timestamp(value: Long): StepikMetricsPostQuery {
        metrics.getMetric()
                .timestamp = value
        return this
    }
    
    fun tags(value: Map<String, String>): StepikMetricsPostQuery {
        metrics.getMetric()
                .tags = value.toMutableMap()
        return this
    }
    
    fun tags(key: String, value: String): StepikMetricsPostQuery {
        metrics.getMetric()
                .addTags(key, value)
        return this
    }
    
    fun tags(key: String, value: Any): StepikMetricsPostQuery {
        metrics.getMetric()
                .addTags(key, value.toString())
        return this
    }
    
    fun data(value: Map<String, Any>): StepikMetricsPostQuery {
        metrics.getMetric()
                .data = value.toMutableMap()
        return this
    }
    
    fun data(key: String, value: Any): StepikMetricsPostQuery {
        metrics.getMetric()
                .addData(key, value)
        return this
    }
    
    override fun getBody(): String {
        return jsonConverter.toJson(metrics, false)
    }
    
    override fun getUrl(): String {
        return "${stepikAction.stepikApiClient.host}/api/metrics"
    }
    
    override fun toString(): String {
        return jsonConverter.toJson(metrics, false)
    }
}
